src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs

```diff
@@ -28,39 +26,57 @@ public InMemoryCryptoProviderCache() : this(new CryptoProviderCacheOptions())
-        /// <param name="tryTakeTimeout">The time used in ms for the timeout interval of the event queue. Defaults to 500 ms.</param>
         internal InMemoryCryptoProviderCache(CryptoProviderCacheOptions cryptoProviderCacheOptions, TaskCreationOptions options, int tryTakeTimeout = 500)
         {
-            if (cryptoProviderCacheOptions == null)
-                throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
-
+            _cryptoProviderCacheOptions = cryptoProviderCacheOptions ?? throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
             if (tryTakeTimeout <= 0)
                 throw LogHelper.LogArgumentException<ArgumentException>(nameof(tryTakeTimeout), $"{nameof(tryTakeTimeout)} must be greater than zero");
 
-            _cryptoProviderCacheOptions = cryptoProviderCacheOptions;
-            _signingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, options, StringComparer.Ordinal, false) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
-            _verifyingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, options, StringComparer.Ordinal, false) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
+            _signingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(
+                cryptoProviderCacheOptions.SizeLimit,
+                options,
+                comparer: StringComparer.Ordinal)
+            {
+                OnItemCompacted = OnSignatureProviderRemoved,
+                OnItemRemoved = DisposeSignatureProvider,
+                OnShouldRemove = ShouldRemoveSignatureProvider,
+            };
+
+            _verifyingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(
+                cryptoProviderCacheOptions.SizeLimit,
+                options,
+                comparer: StringComparer.Ordinal)
+            {
+                OnItemCompacted = OnSignatureProviderRemoved,
+                OnItemRemoved = DisposeSignatureProvider,
+                OnShouldRemove = ShouldRemoveSignatureProvider,
+            };
+        }
+
+
+        internal CryptoProviderFactory CryptoProviderFactory { get; set; }
+
+        private static void DisposeSignatureProvider(SignatureProvider signatureProvider)
+        {
+            signatureProvider.Dispose();
+        }
+
+        private void OnSignatureProviderRemoved(SignatureProvider signatureProvider)
+        {
+            signatureProvider.CryptoProviderCache = null;
+        }
+
+        private static bool ShouldRemoveSignatureProvider(SignatureProvider signatureProvider)
+        {
+            return signatureProvider.CryptoProviderCache == null && signatureProvider.RefCount == 0;
```

Comment:
Should check cache state? Or is this already implied - ShouldRemoveSignatureProvider is called only when provider is cached?

```suggestion
            return signatureProvider.CryptoProviderCache == null && signatureProvider.RefCount == 0 && !signatureProvider.WasCached;
```